complete refactor of auto migration doc (#4183)
temporalio/documentation: docs/cloud/migrate/automated.mdx

  - Production
---

:::tip Support, stability, and dependency info

Automated migration is currently in [Pre-release](/evaluate/development-production-features/release-stages#pre-release).

:::

## Process Overview
Automated migration is designed to provide a zero-downtime, secure means of migrating to Temporal Cloud. This guide outlines the current process

> for transitioning Workflows from a self-hosted setup to one hosted within Temporal Cloud.

Reviewer: ```suggestion
for transitioning Workflows from self-hosted Temporal Server to a namespace hosted within Temporal Cloud.
```

### Namespace migration schedule
When planning a migration it is highly recommended to migrate in the order of least-critical to most-critical Namespace. Ideally, the project

> will begin with Namespaces designated as "testing", where data loss is an acceptable outcome of the testing process. From there, prioritize

Reviewer: ```suggestion
will begin with Namespaces designated as "testing", where downtime is an acceptable outcome of the testing process. From there, prioritize
```

Having "data loss" in the first couple paragraphs is scary.

Migration involves the following phases:

1. Prepare - Initial preparation involves collecting and evaluating key data points from the self-hosted clusters. Collected data will be evaluated
   by Temporal to ensure eligibility for migration.
2. Setup - Once eligibility has been verified, configurations for the self-hosted clusters will be modified to support the migration. All
   cloud-side components will be provisioned and the self-hosted [S2S Proxy](https://hub.docker.com/r/temporalio/s2s-proxy/tags) will be deployed.
3. Test - Once all required components are in place, the migration process will be validated using a test Namespace.
4. Initiate - At the conclusion of a successful testing process, migration of production Namespaces will begin.
5. Finalize - After the Namespace has been transferred to Temporal Cloud and validated, the migration will be finalized.

Please review the [additional notes](#additional-notes) section prior to planning the migration.

## Phase 1: Prepare
In preparation for migrating to Temporal Cloud, the following data points must be collected and provided to Temporal via support ticket.

> - List of Temporal Accounts. If you are new to Temporal Cloud, then the most common recommendation is to create a single Account that contains your Namespaces.

Reviewer: ```suggestion
- List of Temporal Cloud Accounts. If you are new to Temporal Cloud, then the most common recommendation is to create a single Account that contains your Namespaces.
```

- Target Temporal Cloud provider/region per cluster. See the list of [current regions](https://docs.temporal.io/cloud/regions).
- Server Configuration for each cluster.
- For each Namespace:
  - Namespace name and its translated cloud-side name. Since Namespaces in Temporal Cloud follow a [specific format](https://docs.temporal.io/cloud/namespaces#temporal-cloud-namespace-name),
    existing Namespaces must be converted to the new format.
  - Namespace sizing metrics. These are used to properly size cloud-side resources for the migration.

>   - Custom search attributes (if any). Temporal must ensure that your custom search attributes, specifically when using SQL-based visibility store, are compatible with

Reviewer: Link to SA docs?

If you've never enabled Global Namespace, then you may enable it and set the Initial Failover Version to a unique value (between 1 and 100).
Enable by updating the _clusterMetadata_ and the _dcRedirectionPolicy_.
When complete, restart all Temporal services (frontend, history, matching, worker), starting with the frontend.

```yaml
dcRedirectionPolicy:
   policy: "all-apis-forwarding"

clusterMetadata:
enableGlobalNamespace: true # add this
failoverVersionIncrement: CHANGEME # usually 1000000, but coordinate with Temporal. must match cloud-side migration server
masterClusterName: _NO_CHANGE_
currentClusterName: _NO_CHANGE_
clusterInformation:
   _NO_CHANGE_:
      enabled: true
      initialFailoverVersion: CHANGEME # use unique number between 1 and 100 for each server 
      rpcName: _NO_CHANGE_
      rpcAddress: _NO_CHANGE_
```

After all services have restarted, verify the configuration using:

```
temporal operator cluster describe
``` 

The following sample output is expected:
```yaml
"failoverVersionIncrement": "1000000",
"initialFailoverVersion": "the number you picked"
"isGlobalNamespaceEnabled": true
```

### Prepare mTLS certificates

> TLS is used to secure the communications channel for migration. Generate new certificates following the process outlined [here](https://docs.temporal.io/cloud/certificates#issue-certificates).

Reviewer: Use relative link

:::

### Process Overview
Automated migration is designed to provide a zero-downtime, secure means of migrating to Temporal Cloud. This guide outlines the current process
for transitioning Workflows from a self-hosted setup to one hosted within Temporal Cloud.

#### Migration Phases
The migration process is separated into several phases, part of which involves coordinating with Temporal to create necessary cloud-side resources.

Migration involves the following phases:

1. Prepare - Initial preparation involves collecting and evaluating key data points from the self-hosted clusters. Collected data will be evaluated
   by Temporal to ensure eligibility for migration.
2. Setup - Once eligibility has been verified, configurations for the self-hosted clusters will be modified to support the migration. Additionally,
   cloud-side resources will be deployed, and the [S2S Proxy](https://hub.docker.com/r/temporalio/s2s-proxy/tags) will be deployed.
3. Test - Once all required components are in place, the migration process will be tested to ensure full functionality prior to migrating any production Namespaces.

> 4. Initiate - At the conclusion of a successful testing process, migration of production Namespaces may begin.

Author: can you clarify what portion is uncertain?

  a single migration server simultaneously due to cluster name collision. There are 2 available options:
  1. Migrate one server at a time using a single migration server.
  2. Create multiple migration servers (one for each self-hosted server) if you need to migrate all servers simultaneously.
- If you are using multi-cluster replication in your self-hosted setup and have previously failed over Namespaces, then this may impact your eligibility for automated migration.
  Specifically, whenever Global Namespace has been previously enabled the following restrictions apply:
  1. Initial Failover Version must be less than or equal to 1,000,000
  2. Failover Version Increment must be a divisor of 1,000,000
- OSS supports cross-Namespace commands (e.g., parent-child, SignalExternal, CancelExternal) through the `system.enableCrossNamespaceCommands` configuration. This
  configuration is disabled on Temporal Cloud. If cross-Namespace calls exist within Workflow code, they must be updated or removed prior to migration.

#### Additional Notes

> ##### Special dynamic configuration for Version 1.22.x-1.23.x

Reviewer 2: do we mention in the doc that we don't support any version older than 1.22?

  - Production
---

:::tip Support, stability, and dependency info

Automated migration is currently in [Pre-release](/evaluate/development-production-features/release-stages#pre-release).

:::

> ### Process Overview

Reviewer 3: ```suggestion
## Process Overview
```

> ### Process Overview

Reviewer: The page title is h1, so jumping straight to  (h3) skips a heading level. Consider using  for phase headings and  for sub-sections.

Automated migration is designed to provide a zero-downtime, secure means of migrating to Temporal Cloud. This guide outlines the current process
for transitioning Workflows from a self-hosted setup to one hosted within Temporal Cloud.

#### Namespace migration schedule
When planning a migration it is highly recommended to migrate in the order of least-critical to most-critical Namespace. Ideally, the project
will begin with Namespaces designated as "testing", where data loss is an acceptable outcome of the testing process. From there, prioritize
migrations based on order of least potential impact.

![Temporal automated migration components](/img/cloud/migration/ns_migration_schedule.png)

> #### Project phases

Reviewer 3: ```suggestion
### Project phases
```

The migration process is separated into several phases, part of which involves coordinating with Temporal to create necessary cloud-side resources.

Migration involves the following phases:

1. Prepare - Initial preparation involves collecting and evaluating key data points from the self-hosted clusters. Collected data will be evaluated
   by Temporal to ensure eligibility for migration.
2. Setup - Once eligibility has been verified, configurations for the self-hosted clusters will be modified to support the migration. All
   cloud-side components will be provisioned and the self-hosted [S2S Proxy](https://hub.docker.com/r/temporalio/s2s-proxy/tags) will be deployed.
3. Test - Once all required components are in place, the migration process will be validated using a test Namespace.
4. Initiate - At the conclusion of a successful testing process, migration of production Namespaces may begin.
5. Finalize - After the Namespace has been transferred to Temporal Cloud and validated, the migration will be finalized.

> 

Author: claude picky here. can make this change if needed, but the spacing helps me from going blind looking at the md

> #### Limitations

Reviewer 3: ```suggestion
### Limitations
```

  a single migration server simultaneously due to cluster name collision. There are 2 available options:
  1. Migrate one server at a time using a single migration server.
  2. Create multiple migration servers (one for each self-hosted server) if you need to migrate all servers simultaneously.
- If you are using multi-cluster replication in your self-hosted setup and have previously failed over Namespaces, then this may impact your eligibility for automated migration.
  Specifically, whenever Global Namespace has been previously enabled the following restrictions apply:
  1. Initial Failover Version must be less than or equal to 1,000,000
  2. Failover Version Increment must be a divisor of 1,000,000
- OSS supports cross-Namespace commands (e.g., parent-child, SignalExternal, CancelExternal) through the `system.enableCrossNamespaceCommands` configuration. This
  configuration is disabled on Temporal Cloud. All `system.enableCrossNamespaceCommands` must be disabled and code using cross-Namespace calls must be updated or removed prior to migration.

#### Additional notes

> ##### Special dynamic configuration for Version 1.22.x-1.23.x

Reviewer 3: ```suggestion
#### Special dynamic configuration for Version 1.22.x-1.23.x
```

Temporal versions 1.22.x and 1.23.x include support for stream-based replication, but it is disabled by default. Since those releases, stream-based replication has been validated
as more reliable than the poll-based replication that remained the default in 1.22 and 1.23. 

When preparing for an S2C migration on these versions, configure the following dynamic settings to enable stream-based replication:

```yaml
history.enableReplicationStream:
  - value: true
```

Enabling this configuration will require a restart of your history pods. 

> ### Phase 1: Prepare

Reviewer 3: ```suggestion
## Phase 1: Prepare
```

```
List of Accounts:

<for each cluster>
Cluster Name:
Target Cloud/Region:
Cluster Configuration:

List of Namespaces:
   <for each Namespace>
   Namespace Name:
   Translated Name:
   Metrics:
   Custom Search Attributes:
   <end>
<end>
```

### Phase 2: Setup
Once the migration has been approved, the next step is to prepare both the self-hosted clusters and Temporal Cloud resources for the migration.

:::warning

Proceed only when your request has been approved by Temporal.

:::

#### Prepare Temporal Cloud Namespace
For each Namespace to be migrated, create an empty Namespace in Temporal Cloud to serve as the migration target. Since migration cannot proceed into a Namespace that's already
in use, the Namespace should remain empty (no Workflows). Apply any required custom search attributes.

Be sure to adjust the [rate limits](https://docs.temporal.io/cloud/capacity-modes) of your Namespace as needed and create any custom search attributes that are required.

If you are new to Temporal Cloud, consider your desired connectiviy path to your Namespace. You may connect over the public internet or via 
[private connectivity](https://docs.temporal.io/cloud/connectivity).

> #### Modify cluster configuration

Reviewer 3: ```suggestion
### Modify cluster configuration
```

:::

Adjust the maximum connection keepalive time to match the setting in cloud.
```yaml
frontend.keepAliveMaxConnectionAge:
- value: "2h"
```

If you have previously enabled [Global Namespace](https://docs.temporal.io/global-namespace), the Failover Version Increment must be approved by Temporal.
This is **critical**, since the value may only be set once and cannot change.

If you've never enabled Global Namespace, then you may enable it and set the Initial Failover Version to a unique value (between 1 and 100).
Enable by updating the _clusterMetadata_ and the _dcRedirectionPolicy_.
When complete, restart all Temporal services (frontend, history, matching, worker), starting with the frontend.

```yaml
dcRedirectionPolicy:
   policy: "all-apis-forwarding"

clusterMetadata:
enableGlobalNamespace: true # add this
failoverVersionIncrement: 1000000 # coordinate with Temporal. must match cloud-side migration server
masterClusterName: _NO_CHANGE_
currentClusterName: _NO_CHANGE_
clusterInformation:
   _NO_CHANGE_:
      enabled: true
      initialFailoverVersion: [1,100] # use unique number between 1 and 100 for each server 
      rpcName: _NO_CHANGE_

>       rpcAddress: _NO_CHANGE_

Author: added a CHANGEME placeholder to avoid situation where this config is applied an some example default value is used by mistake

Use the following procedure to deploy the proxy:
1. Obtain the latest Docker image from the [temporalio/s2s-proxy repository](https://hub.docker.com/r/temporalio/s2s-proxy/tags).
2. Gather the mTLS certs generated in the previous step.
3. Deploy 1 replica of the s2s-proxy (minimum 4 CPU and 512mb memory). For Kubernetes users, use this [helm chart example](https://github.com/temporalio/s2s-proxy/blob/main/charts/s2s-proxy/README.md)
   as a reference. See the [example](https://github.com/temporalio/s2s-proxy/blob/main/charts/s2s-proxy/example.yaml) configuration file as a reference.
4. Test access using the command below. It should display the information of the migration server.

```
temporal operator cluster describe --address {the-outbound-external-address-of-your-proxy}
```

Once connectivity has been verified, notify Temporal so that connectivity may be validated from the cloud-side proxy.

### Phase 3: Testing
Testing should proceed using either a newly created Namespace or else one that is considered to be non-production. In either case, test Workflows
should be utilized. It is ideal to have a mix of completed, active, and new Workflows to use during testing.

> Testing will follow the same process as the production migration. Use the steps outlined in the remaining phases to perform testing.

Reviewer: This phase could use more detail: what does a successful test look like? And if a test migration fails, what are the cleanup/recovery steps before retrying?

Author: added steps + success criteria

### Phase 4: Initiate
The sections below outline the process for initiating the migration.

#### Migration start
Temporal will intiate the migration. During this process, the self-hosted Namespace remains active while the cloud Namespace becomes passive.
Workflows are replicated from the self-hosted Namespace to the cloud Namespace. Once cloud Namespace has fully synced with self-hosted Namespace,
migration is ready for handover.

The following [command](https://pkg.go.dev/github.com/temporalio/tcld#readme-start-a-migration) is used to start the migration:

```
tcld migration start --endpoint-id <endpoint-id> --source-namespace <source-namespace> --target-namespace <target-namespace>
```

#### Monitor
During the initial sync, it is important to monitor the overall process to ensure progress is being made. While Temporal will monitor from the cloud-side,
progress may also be monitored from the self-hosted side using the _replication_stream_stuck_ metric.

> Additionally, there are multiple [metrics](https://github.com/temporalio/s2s-proxy/blob/main/metrics/prometheus_defs.go) available on the S2S proxy itself.

Author: removed link. will update with specific callouts for a few metrics as a later update

- Monitor general Workflow metrics (schedule to start latency, start v.s. completion rate, sync match rate, etc).
- Learn how [capacity management](https://docs.temporal.io/cloud/capacity-modes) works in Temporal Cloud.
- Plan for a worker tuning session - performance change between Temporal Cloud v.s. self-hosted cluster, which could lead to unexpected symptoms and optimizations.
- Know how to reach out to your Temporal Solutions Architect (SA) and Account Executive (AE) for assistance.

#### Confirm complete
Once a Namespace has been transferred to the cloud and validated, the migration will be completed. Note that this step is final and may not be undone. Once performed, 
Workflow replication from the cloud Namespace to the self-hosted server is halted. 

The following [command](https://pkg.go.dev/github.com/temporalio/tcld#readme-confirm-a-migration) is used to confirm the migration:

```
tcld migration confirm --id <migration-id>
```

or to [abort](https://pkg.go.dev/github.com/temporalio/tcld#readme-abort-a-migration) and roll-back changes without impacting your Workflows, if needed:

```
tcld migration abort --id <migration-id>
```

> #### Transfer clients to cloud

Reviewer: This section is referenced from Handover ("See the instructions below") but appears *after* "Confirm complete". Workers need to be transferred before confirming. Consider moving this section before "Confirm complete", or at minimum make the sequencing explicit.

Deploy two sets of Temporal clients: one pointing to your Temporal server and one to the Cloud Namespace endpoint. This is the recommended option since your Workflows will
continue to make progress during the handover, even if your cloud Temporal client is unable to access the cloud (due to misconfiguration, for example). The process is as follows:
1. Direct your cloud-based Temporal clients to the cloud Namespace endpoint. Initially, these clients will connect and send Poll requests but will not receive any tasks.
2. Start migration. Your self-hosted Namespace is active while your cloud Namespace is passive (or standby). Your cloud Temporal clients can begin receiving tasks, but all requests
   from cloud clients to the Cloud Namespace will automatically forward from the cloud to your self-hosted server.
3. Hand over Namespace to the cloud. Your cloud Namespace becomes active and your self-hosted Namespace becomes passive. All requests from your self-hosted Temporal clients will
   automatically forward from your server to the cloud.
4. Complete migration. Your self-hosted Temporal clients will no longer receive any tasks from your server, allowing you to stop these clients.

> ##### Option 2 

Reviewer 3: ```suggestion
#### Option 2 
```

4. Confirm migration completion.

### Frequently asked questions

**What Workflows are migrated by default?**

Open workflows are always migrated. You may specify a date range for closed Workflows. Limiting the time range for closed Workflows will greatly reduce the
amount of time it takes to migrate a Namespace. Your cloud-side Namespace must be configured with your desired retention period prior to starting the migration.

**I have a long retention period for my Workflows. Is this compatible with Temporal Cloud?**

Occasionally, self-hosted [retention periods](https://docs.temporal.io/temporal-service/temporal-server#retention-period) are in excess of what is 
[supported](https://docs.temporal.io/cloud/limits#default-retention-period) in Temporal Cloud.
In these cases it is recommended to utilize [archival](https://docs.temporal.io/temporal-service/archival) to store closed Workflows that cannot be
migrated. In general, archival is recommended over large retention periods since the extra data can stress the persistence layer of the system.

> ### How to gather self-hosted metrics

Reviewer 3: ```suggestion
## How to gather self-hosted metrics
```

Enabling this configuration will require a restart of your history pods. 

## Frequently asked questions

**Why does it matter if custom search attributes are used?**

Custom search attributes must be mapped to the Namespace in Temporal Cloud. This process differs depending on the type of data store used for visibility.

**What Workflows are migrated by default?**

Open workflows are always migrated. You may specify a date range for closed Workflows. Limiting the time range for closed Workflows will greatly reduce the
amount of time it takes to migrate a Namespace. Your cloud-side Namespace must be configured with your desired retention period prior to starting the migration.

**I have a long retention period for my Workflows. Is this compatible with Temporal Cloud?**

Occasionally, self-hosted [retention periods](https://docs.temporal.io/temporal-service/temporal-server#retention-period) are in excess of what is 

> [supported](https://docs.temporal.io/cloud/limits#default-retention-period) in Temporal Cloud.

Reviewer: Use relative link
